mmv1/products/firebaseappcheck/DebugToken.yaml

```diff
@@ -0,0 +1,93 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+!ruby/object:Api::Resource
+name: "DebugToken"
+base_url: projects/{{project}}/apps/{{app_id}}/debugTokens
+self_link: projects/{{project}}/apps/{{app_id}}/debugTokens/{{debug_token_id}}
+update_verb: :PATCH
+update_mask: true
+description: |
+  A debug token is a secret used during the development or integration testing of
+  an app. It essentially allows the development or integration testing to bypass
+  app attestation while still allowing App Check to enforce protection on supported
+  production Firebase services.
+references: !ruby/object:Api::Resource::ReferenceLinks
+  guides:
+    "Official Documentation": "https://firebase.google.com/docs/app-check"
+  api: "https://firebase.google.com/docs/reference/appcheck/rest/v1/projects.apps.debugTokens"
+import_format:
+  [
+    "projects/{{project}}/apps/{{app_id}}/debugTokens/{{debug_token_id}}",
+    "{{project}}/{{app_id}}/{{debug_token_id}}",
+    "{{app_id}}/{{debug_token_id}}",
+  ]
+examples:
+  - !ruby/object:Provider::Terraform::Examples
+    name: "firebase_app_check_debug_token_basic"
+    # Need the time_sleep resource
+    pull_external: true
+    primary_resource_id: "default"
+    vars:
+      display_name: "Debug Token"
+      token: "00000000-AAAA-BBBB-CCCC-000000000000"
+    test_vars_overrides:
+      # Don't add prefix. This string is a random UUID4
+      token: "\"5E728315-E121-467F-BCA1-1FE71130BB98\""
+    test_env_vars:
+      project_id: :PROJECT_NAME
+parameters:
+  - !ruby/object:Api::Type::String
+    name: app_id
+    description: |
+      The ID of a
+      [Web App](https://firebase.google.com/docs/reference/firebase-management/rest/v1beta1/projects.webApps#WebApp.FIELDS.app_id),
+      [Apple App](https://firebase.google.com/docs/reference/firebase-management/rest/v1beta1/projects.iosApps#IosApp.FIELDS.app_id),
+      or [Android App](https://firebase.google.com/docs/reference/firebase-management/rest/v1beta1/projects.androidApps#AndroidApp.FIELDS.app_id)
+    required: true
+    immutable: true
+    url_param_only: true
+properties:
+  - !ruby/object:Api::Type::String
+    name: name
+    description: |
+      The fully-qualified resource name of the debug token.
+    output: true
+    pattern: projects/{{project}}/apps/{{app_id}}/debugTokens/{{debug_token_id}}
+    custom_flatten: 'templates/terraform/custom_flatten/firebase_app_check_debug_token.go.erb'
+  - !ruby/object:Api::Type::String
+    name: debug_token_id
```

Comment:
I think this should be url_param_only instead of output. That also implies ignore_read

Reply:
It should also be required.

Reply:
`debug_token_id` is generated by the server upon creation. The user is not expected to supply it. I guess this means it should be output only? This field is still needed to maintain the structure of the `self_link` for import & reading.

Reply:
Ah I thought the user could set it.

In this case I think we should just use `name` as the output-only field, remove debug_token_id, set self_link and the resource ID to {{name}} and remove the custom flattener

Reply:
How can I still support importing with all 3 formats?
```
[
    "projects/{{project}}/apps/{{app_id}}/debugTokens/{{debug_token_id}}",
    "{{project}}/{{app_id}}/{{debug_token_id}}",
    "{{app_id}}/{{debug_token_id}}",
]
```

Reply:
Is it important to support all 3 formats?

Reply:
Just a `{{name}}` import format has caused confusion before. See https://github.com/hashicorp/terraform-provider-google/issues/7907

Thus, I'm learning towards supporting it if able.

Reply:
Ok, how about just using {{name}} to represent the identifier that appears after debugTokens/

You can remove all instances of debug_token_id, replace with name, and use the standard 'templates/terraform/custom_flatten/name_from_self_link.erb' custom flatten

Reply:
I've also seen over resources do this. While this is probably the path with least custom code, I'm concerned about possible user confusion around ID and name. The underlying API neatly follows [AIP-122](https://google.aip.dev/122) i.e.

```
resource "google_firebase_app_check_debug_token" "example" {
  name = "server_generated_debug_token_id"
}
```
The AIP-122 Resource name for this Terraform resource is `projects/{{project}}/apps/{{app_id}}/debugTokens/server_generated_debug_token_id`, but `google_firebase_app_check_debug_token.example.name` gives “server_generated_debug_token_id”. In order to retrieve the AIP-122 resource name, users need to use `google_pubsub_schema.example.id`. As you can see, Terraform's `ID` and `name` are swapped when compared with AIP-122's `resource ID` and `resource name`.

I'm fine with this solution. Is there any prior discussion in the team for this phenomenon? I expect more and more modern APIs to follow AIP-122 so the issue only becomes bigger. An official guidance would be appreciated here.

Reply:
As you indicated in the previous message, often users are confused by the long-form name 😉

It's fairly common in these Terraform providers for the `name` field to mean just the short-form ending of the self link, and the `.id` field is the long-form. So in this case I think it's sticking to precedent to continue using name as such. It might actually be worth calling it something different to identify that it's a server-set name rather than user-settable (you could go back to the original debug_token_id). In this case you could use `api_name: name` and the shared custom_flatten, I just have a problem with using a custom flatten to read from a different field.

The AIPs are not commonly known by users, many APIs (like this one!) don't follow them all too well, and they aren't meant to be guidelines on how client surfaces represent APIs

Reply:
YES, this is what I'm looking for. I didn't know about the `api_name` field, but it seems perfect for this use case. I was able to get it working with the additional `identity` field. I'm very happy with how this turns out.